### PR TITLE
Fix bug in copying private key

### PR DESCRIPTION
**Pull request by boscohyun**
SSIA

### CODE REVIEW

File: src/renderer/views/account/CreateAccountView.tsx [hunk 1]
```diff
@@ -34,11 +34,20 @@ const CreateAccountView = observer(
...
+      if (
+        typeof unprotectedPrivateKey[0] !== "string" ||
+        unprotectedPrivateKey[0] === ""
+      ) {
+        // FIXME: Show a new error page or retry page to complete the account creation.
+        console.error("Failed to unprotect private key");
+        return;
+      }
+
+      const privateKey: string = unprotectedPrivateKey[0];
```
Comment (moreal) [inline src/renderer/views/account/CreateAccountView.tsx:50]: How about using [`Destructuring assignment`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)? And it passes also error object like Go does. So how about checking error with it? Like...

```suggestion
      const [privateKey, error]: [
        string | undefined,
        Error | undefined
      ] = ipcRenderer.sendSync("unprotect-private-key", address, password);
      if (
        error !== undefinded ||
        typeof unprotectedPrivateKey[0] !== "string" ||
        unprotectedPrivateKey[0] === ""
      ) {
        // FIXME: Show a new error page or retry page to complete the account creation.
        console.error("Failed to unprotect private key");
        return;
      }
```

**Review by area363 — APPROVED** [review]

File: src/renderer/views/account/CreateAccountView.tsx (hunk 1 above)
Comment (boscohyun) [reply src/renderer/views/account/CreateAccountView.tsx]: @moreal I fixed it!

File: src/renderer/views/account/CreateAccountView.tsx [hunk 2]
```diff
@@ -34,11 +34,15 @@ const CreateAccountView = observer(
         password
       );
 
-      const privateKey: PrivateKey = ipcRenderer.sendSync(
-        "unprotect-private-key",
-        address,
-        password
-      );
+      const [privateKey, passphraseOrError]: [
```
Comment (moreal) [inline src/renderer/views/account/CreateAccountView.tsx:37]: It seems the second element of the return value has been never `passphrase`.

https://github.com/planetarium/9c-launcher/blob/78c0ee02eca6a18df6af74d1dcee1b0cf3979bb4/src/main/main.ts#L539-L562

So it should rename into `error`.

Comment (boscohyun) [reply src/renderer/views/account/CreateAccountView.tsx:37]: Oh my mistake. The js(ts) convention is confusing to me.😢

Comment (boscohyun) [reply src/renderer/views/account/CreateAccountView.tsx:37]: The second element can be `{}`, `undefined`, `error: any`. How can I naming it? Hm... Just `any` or `_` I think.

Comment (boscohyun) [reply src/renderer/views/account/CreateAccountView.tsx:37]: Here is [a interesting article](https://ntgard.medium.com/tuples-in-javascript-cd33321e5277)
```
const [, , thisOne, , thatOne] = getTuple();
const [_1, _2, thisOne, _3, thatOne] = getTuple();
```

Comment (moreal) [reply src/renderer/views/account/CreateAccountView.tsx]: Then, I suggest to ignore the second element like the post you found `const [privateKey] = ipc...` if you will check the error with only `privateKey` state without `error`. But I suggest also to check with `error`.

And I have a question, what the case the second element became `{}`?

Comment (boscohyun) [reply src/renderer/views/account/CreateAccountView.tsx]: The second element became `{}` on line 547.

**Review by moreal — APPROVED** [review]

File: src/renderer/views/account/CreateAccountView.tsx (hunk 2 above)
Comment (moreal) [reply src/renderer/views/account/CreateAccountView.tsx]: I see 👍🏻

Comment (boscohyun) [reply src/renderer/views/account/CreateAccountView.tsx]: The `{}` on line 547 can change to `error: string` that contains error message. How's it? @moreal 
And store the second element as `error: any | undefined` type and handles the error.

Comment (moreal) [reply src/renderer/views/account/CreateAccountView.tsx]: > And store the second element as `error: any | undefined` type and handles the error.

`any` contains `undefined` so there is no reason to use `any | undefined`.

> The `{}` on line 547 can change to `error: string` that contains error message

Could you describe more about the change you mean? If you mean making it return other meaningful value instead `{}`, it seems better to return `Error` instead `string`, to suit for `Error` concepts.

File: src/errors/UndefinedProtectedPrivateKeyError.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class UndefinedProtectedPrivateKeyError extends Error {
+  constructor(msg: string) {
+    super(msg);
+    Object.setPrototypeOf(this, UndefinedProtectedPrivateKeyError.prototype);
+  }
+}
+
+export default UndefinedProtectedPrivateKeyError;
```
Comment (riemannulus) [inline src/errors/UndefinedProtectedPrivateKeyError.ts:8]: Can we use function component instead class component? https://ko.reactjs.org/docs/components-and-props.html#function-and-class-components

Comment (moreal) [reply src/errors/UndefinedProtectedPrivateKeyError.ts:8]: It seems an inheritance of `Error` class, not `React.Component`.

Comment (riemannulus) [reply src/errors/UndefinedProtectedPrivateKeyError.ts:8]: Oh yes 😅  I misreading class name lol

**Review by riemannulus — APPROVED** [review]

File: src/errors/UndefinedProtectedPrivateKeyError.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class UndefinedProtectedPrivateKeyError extends Error {
+  constructor(msg: string) {
+    super(msg);
+    Object.setPrototypeOf(this, UndefinedProtectedPrivateKeyError.prototype);
```
Comment (moreal) [inline src/errors/UndefinedProtectedPrivateKeyError.ts:4]: It is a question, why does it call [`Object.setPrototypeOf`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf)? (P.S. I read [Mozilla Custom Error Types section](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#custom_error_types).)

Comment (boscohyun) [reply src/errors/UndefinedProtectedPrivateKeyError.ts:4]: I'm curious too.🤔

Comment (boscohyun) [reply src/errors/UndefinedProtectedPrivateKeyError.ts:4]: I agree.

**Pull request merged**